src/dune_pkg/fiber_cache.ml

```diff
@@ -0,0 +1,70 @@
+open! Import
+
+type ('k, 'v) t =
+  { (* The cache stores results so that if an exception was raised while
+       computing a value, concurrent accesses to the cache can find out. *)
+    table : ('k, ('v, Exn_with_backtrace.t list) result Fiber.Ivar.t) Table.t
+      (* This module is stored so it can be reused to create a table from the cache. *)
+  ; key_module : (module Table.Key with type t = 'k)
+  }
+
+let create key_module = { table = Table.create key_module 1; key_module }
+
+(** Compute the value to associate with a key by calling the given function,
+    replacing the current value in the cache (if any). If [f] raises then the
+    ivar is removed from the cache and this function will also raise the same
+    exceptions. Note that if [find_or_add] is called on the same key while the
+    ivar is in the table, and [f] raises exceptions, that call to [find_or_add]
+    will also raise the same exceptions. *)
+let compute_value t key ~f =
+  let open Fiber.O in
+  (* Add the empty ivar to the table before calling [f] so that if the key
+     requested again while [f] is running then the caller will wait on the
+     ivar that will eventually contain the result. *)
+  let ivar = Fiber.Ivar.create () in
+  Table.set t.table key ivar;
+  let* value_result = Fiber.collect_errors f in
+  let* () = Fiber.Ivar.fill ivar value_result in
+  match value_result with
+  | Ok value -> Fiber.return value
+  | Error errors ->
+    (* Remove the ivar from the table so that future calls to [find_or_add]
+       attempt to recompute the value. *)
+    Table.remove t.table key;
```

Comment:
I think this behavior is weird b/c it is non deterministic. We should always either recompute on error, or we raise the same errors.

Reply:
Fair enough. Updated so that it always raises. I thought about having it attempt to recompute but it was getting complicated and in most cases I would expect the recomputation to fail in the same way.